docs/architecture/pk-4.md

```diff
@@ -0,0 +1,394 @@
+
+* `mbedtls_pk_set_algorithm()` — but as noted above this is not future-proof.
+* `mbedtls_pk_import_into_psa()` then `mbedtls_pk_copy_from_psa()`. Inefficient but ok. If that's the intended way to do it, we need to document it clearly, and maybe we should remove `mbedtls_pk_set_algorithm()`?
+* A new function `mbedtls_pk_copy()` allowing a policy change?
+
+### Resource management
+
+#### Access the underlying PSA key of a PK context
+
+Should we provide a function to access the underlying PSA key of a PSA context, if there is one?
+
+This would be new work, and does not seem to be needed at the moment. If the PK context was created from a PSA key, the application might as well use the original PSA key. If the PK context was created by parsing, `mbedtls_pk_import_into_psa()` works, and does not require a special case if the PK context does not have an underlying PSA key.
+
+If we add this in the future, it will be considerably easier if all PK contexts have an underlying PSA key, or at least all PK contexts containing a private key have an underlying PSA key.
+
+## Later tasks
+
+The tasks described in this section do not need to be done before the 1.0/4.0 release. This section is incomplete.
+
+### Missing functionality
+
+#### Remove limitations of `mbedtls_pk_wrap_psa()`
+
+Remove the limitations of wrapped keys:
+
+* Implement verify for wrapped RSA keys.
+* Implement check-pair for wrapped RSA keys.
+
+### Migrate library and test code
+
+#### Update terminology from “opaque” to “wrapped”
+
+Update the library code to change the obsolete terminology “opaque” (as in e.g. `MBEDTLS_PK_OPAQUE`) to “wrapped”.
+
+Originally, PK contexts could wrap a PSA key in order to support non-exportable keys kept in a secure partition, using a PSA implementation with client-server isolation. Thus the useful purpose of wrapping a PSA key was to make the key opaque, hence the name. In TF-PSA-Crypto, we plan to evolve to making PK keys contain a PSA key in more situations. What will matter is whether the underlying PSA key is owned by the PK context (and destroyed when the context is freed), or not (thus surviving when the context is freed). A key owned by the PK context could still potentially be opaque. Hence the distinction should be between _wrapped_ and _owned_ PSA keys.
+
+#### Retire `MBEDTLS_PK_ECDSA`
+
+It is no longer possible to construct a PK object with the legacy type `MBEDTLS_PK_ECDSA`. So get rid of the code that handles it, and remove `MBEDTLS_PK_ECDSA`.
+
+#### Replace all uses of `mbedtls_pk_type_t`, `mbedtls_pk_get_type()` and `mbedtls_pk_can_do()` in X.509 library and test code
+
+#### Replace all uses of `mbedtls_pk_type_t`, `mbedtls_pk_get_type()` and `mbedtls_pk_can_do()` in TLS library and test code
+
+#### Get rid of `mbedtls_pk_can_do()`
```

Comment:
I thought this function was being removed as part of the main effort?

Do you mean one of the following?:
* Get rid of `mbedtls_pk_can_do_ext()`.
* `mbedtls_pk_can_do()` has been privatised, entirely delete it (remove from internal use as well).

Reply:
We can't get rid of `mbedtls_pk_can_do` since it's still heavily used, mostly in X.509. All we can do is prevent external users from using it, then remove our own uses when we have time, which is after the 1.0 release.